File: apps/Applications/ios/Local For Vocal/Local For Vocal/Sources/Views/Pages/Grocery/GroceryCategoryPageView.swift
import SwiftUI

struct GroceryCategoryPageView: View {
    @State private var categoryGroups: [CategoryGroup] = []
    @State private var isLoading = true

    // Hardcoded ID for the main Grocery category as per requirement
    private let GROCERY_PARENT_ID = "697095953758a7d8f76fa88c"

    // Grid Configuration
    private let columns = [
        GridItem(.flexible(), spacing: 16),
        GridItem(.flexible(), spacing: 16),
        GridItem(.flexible(), spacing: 16),
        GridItem(.flexible(), spacing: 16),
    ]

    var body: some View {
        VStack(spacing: 0) {
            // Fixed Header
            HStack {
                Text("All Categories")
                    .font(.system(size: 20, weight: .bold))
                    .foregroundColor(Color(hex: "#374151"))
                Spacer()
            }
            .padding(.horizontal, 16)
            .padding(.top, 10)  // Adjust for status bar overlap if needed, though background handles visual
            .padding(.bottom, 12)
            .background(Color(hex: "#FFF8E7").ignoresSafeArea(edges: .top))
            .shadow(color: Color.black.opacity(0.05), radius: 2, x: 0, y: 1)
            .zIndex(1)  // Ensure it stays on top

            if isLoading {
                VStack {
                    ProgressView()
                    Text("Loading categories...")
                        .font(.system(size: 14))
                        .foregroundColor(.gray)
                        .padding(.top, 8)
                }
                .frame(maxWidth: .infinity, maxHeight: .infinity)
            } else {
                ScrollView(showsIndicators: false) {
                    LazyVStack(alignment: .leading, spacing: 24) {
                        ForEach(categoryGroups, id: \.name) { group in
                            VStack(alignment: .leading, spacing: 16) {
                                // Group Header
                                Text(group.name)
                                    .font(.system(size: 18, weight: .bold))
                                    .foregroundColor(Color(hex: "#111827"))
                                    .padding(.horizontal, 16)

                                // Group Items Grid
                                LazyVGrid(columns: columns, spacing: 20) {
                                    ForEach(group.categories) { category in
                                        NavigationLink(destination: destinationView(for: category))
                                        {
                                            CategoryGridItem(category: category)
                                        }
                                        .buttonStyle(PlainButtonStyle())
                                    }
                                }
                                .padding(.horizontal, 16)
                            }
                        }
                    }
                    .padding(.vertical, 20)
                }
            }
        }
        .background(Color.white)
        .onAppear {
            loadCategories()
        }
    }

    private func destinationView(for category: CategoryModel) -> some View {
        // Navigate to GroceryListingView with the selected subcategory
        GroceryListingView(
            categoryId: GROCERY_PARENT_ID,
            categoryName: "Grocery",
            initialSubCategoryId: category.id
        )
    }

    private func loadCategories() {
        Task {
            do {
                let allCategories = try await APIService.shared.fetchCategories()

                // 1. Find the parent Grocery category to get the group order
                let parentCategory = allCategories.first { $0.id == GROCERY_PARENT_ID }
                let groupOrder = parentCategory?.subCategoryGroupOrder ?? []

                // 2. Filter for subcategories of Grocery
                let grocerySubcategories = allCategories.filter {
                    $0.parentCategory == GROCERY_PARENT_ID
                }

                // 3. Group by 'group' property
                let grouped = Dictionary(grouping: grocerySubcategories) { $0.group ?? "Other" }

                // 4. Sort groups based on groupOrder
                var sortedGroups: [CategoryGroup] = []

                // First add groups that are in the order list
                for groupName in groupOrder {
                    if let cats = grouped[groupName] {
                        sortedGroups.append(CategoryGroup(name: groupName, categories: cats))
                    }
                }

                // Then add any remaining groups (e.g. "Other" or new groups not in order list)
                // Filter out keys already added
                let remainingKeys = grouped.keys.filter { !groupOrder.contains($0) }.sorted()
                for groupName in remainingKeys {
                    if let cats = grouped[groupName] {
                        sortedGroups.append(CategoryGroup(name: groupName, categories: cats))
                    }
                }

                await MainActor.run {
                    self.categoryGroups = sortedGroups
                    self.isLoading = false
                }

            } catch {
                AppLogger.error("Failed to load grocery categories: \(error)")
                await MainActor.run {
                    self.isLoading = false
                }
            }
        }
    }
}

// MARK: - Helper Models & Views

struct CategoryGroup {
    let name: String
    let categories: [CategoryModel]
}

struct CategoryGridItem: View {
    let category: CategoryModel

    var body: some View {
        VStack(spacing: 8) {
            // Image Container
            ZStack {
                RoundedRectangle(cornerRadius: 16)
                    .fill(Color(hex: "#F3F4F6"))  // Light gray background
                    .aspectRatio(1, contentMode: .fit)

                if let imageUrl = category.image,
                    let url = URL(
                        string: imageUrl.hasPrefix("http")
                            ? imageUrl : "\(APIService.shared.imageHost)\(imageUrl)")
                {
                    AsyncImage(url: url) { image in
                        image
                            .resizable()
                            .aspectRatio(contentMode: .fit)
                            .padding(8)  // Padding inside the box
                    } placeholder: {
                        ProgressView()
                    }
                } else if let icon = category.icon {
                    // Fallback to icon if image is missing
                    Image(systemName: icon)
                        .font(.system(size: 24))
                        .foregroundColor(.gray)
                }
            }
            .frame(maxWidth: .infinity)

            // Category Name
            Text(category.name)
                .font(.system(size: 12, weight: .medium))
                .foregroundColor(Color(hex: "#374151"))
                .multilineTextAlignment(.center)
                .lineLimit(2)
                .fixedSize(horizontal: false, vertical: true)
        }
    }
}

File: apps/Applications/android/app/src/main/java/com/ecommerceearn/app/ui/pages/GroceryCategoryPageView.kt
package com.ecommerceearn.app.ui.pages

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.zIndex
import coil.compose.AsyncImage
import com.ecommerceearn.app.data.model.Category
import com.ecommerceearn.app.data.remote.NetworkClient
import com.ecommerceearn.app.utils.AppLogger
import kotlinx.coroutines.launch

data class CategoryGroup(
    val name: String,
    val categories: List<Category>
)

@Composable
fun GroceryCategoryPageView() {
    val GROCERY_PARENT_ID = "697095953758a7d8f76fa88c"
    var categoryGroups by remember { mutableStateOf<List<CategoryGroup>>(emptyList()) }
    var isLoading by remember { mutableStateOf(true) }
    val coroutineScope = rememberCoroutineScope()

    LaunchedEffect(Unit) {
        coroutineScope.launch {
            try {
                val allCategories = com.ecommerceearn.app.data.repository.CategoryRepository.getCategories()
                
                // 1. Find the parent Grocery category to get the group order
                val parentCategory = allCategories.find { it._id == GROCERY_PARENT_ID }
                val groupOrder = parentCategory?.subCategoryGroupOrder ?: emptyList()

                // 2. Filter for subcategories of Grocery
                val grocerySubcategories = allCategories.filter { it.parentCategory == GROCERY_PARENT_ID }

                // 3. Group by 'group' property
                val grouped = grocerySubcategories.groupBy { it.group ?: "Other" }

                // 4. Sort groups based on groupOrder
                val sortedGroups = mutableListOf<CategoryGroup>()

                for (groupName in groupOrder) {
                    grouped[groupName]?.let { cats ->
                        sortedGroups.add(CategoryGroup(name = groupName, categories = cats))
                    }
                }

                val remainingKeys = grouped.keys.filter { !groupOrder.contains(it) }.sorted()
                for (groupName in remainingKeys) {
                    grouped[groupName]?.let { cats ->
                        sortedGroups.add(CategoryGroup(name = groupName, categories = cats))
                    }
                }

                categoryGroups = sortedGroups
            } catch (e: Exception) {
                AppLogger.error("Failed to load grocery categories: ${e.message}")
            } finally {
                isLoading = false
            }
        }
    }

    Column(modifier = Modifier.fillMaxSize().background(Color.White)) {
        // Fixed Header
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .zIndex(1f)
                .shadow(elevation = 2.dp)
                .background(Color(0xFFFFF8E7))
                .statusBarsPadding()
                .padding(horizontal = 16.dp, vertical = 12.dp) // adjusted for Android
        ) {
            Text(
                text = "All Categories",
                fontSize = 20.sp,
                fontWeight = FontWeight.Bold,
                color = Color(0xFF374151)
            )
        }

        if (isLoading) {
            Box(modifier = Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
                Column(horizontalAlignment = Alignment.CenterHorizontally) {
                    CircularProgressIndicator(color = Color(0xFF2874F0))
                    Spacer(modifier = Modifier.height(8.dp))
                    Text("Loading categories...", fontSize = 14.sp, color = Color.Gray)
                }
            }
        } else {
            LazyVerticalGrid(
                columns = GridCells.Fixed(4),
                modifier = Modifier.fillMaxSize(),
                contentPadding = PaddingValues(start = 16.dp, end = 16.dp, top = 20.dp, bottom = 120.dp),
                horizontalArrangement = Arrangement.spacedBy(16.dp),
                verticalArrangement = Arrangement.spacedBy(20.dp)
            ) {
                categoryGroups.forEach { group ->
                    item(span = { androidx.compose.foundation.lazy.grid.GridItemSpan(maxLineSpan) }) {
                        Text(
                            text = group.name,
                            fontSize = 18.sp,
                            fontWeight = FontWeight.Bold,
                            color = Color(0xFF111827),
                            modifier = Modifier.padding(top = 12.dp, bottom = 8.dp)
                        )
                    }

                    items(group.categories) { category ->
                        CategoryGridItem(category = category)
                    }
                }
            }
        }
    }
}

@Composable
fun CategoryGridItem(category: Category) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clickable {
                // Future navigation logic mapping to List Page
            },
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .aspectRatio(1f)
                .background(Color(0xFFF3F4F6), RoundedCornerShape(16.dp)),
            contentAlignment = Alignment.Center
        ) {
            if (!category.image.isNullOrEmpty()) {
                val imageUrl = if (category.image.startsWith("http")) category.image else "https://api.lfvs.in/${category.image.removePrefix("/")}"
                AsyncImage(
                    model = imageUrl,
                    contentDescription = category.name,
                    contentScale = ContentScale.Fit,
                    modifier = Modifier
                        .fillMaxSize()
                        .padding(8.dp)
                )
            } else {
                Text(
                    text = category.icon ?: "📂",
                    fontSize = 24.sp,
                    color = Color.Gray
                )
            }
        }
        Spacer(modifier = Modifier.height(8.dp))
        Text(
            text = category.name,
            fontSize = 12.sp,
            fontWeight = FontWeight.Medium,
            color = Color(0xFF374151),
            textAlign = TextAlign.Center,
            maxLines = 2,
            overflow = TextOverflow.Ellipsis
        )
    }
}
